sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -450,16 +451,32 @@ case class Lower(child: Expression)
   override protected def withNewChildInternal(newChild: Expression): Lower = copy(child = newChild)
 }
 
-/** A base trait for functions that compare two strings, returning a boolean. */
-abstract class StringPredicate extends BinaryExpression
+/** A base trait for functions that compare two strings or binaries, returning a boolean. */
+abstract class BinaryPredicate extends BinaryExpression
   with Predicate with ImplicitCastInputTypes with NullIntolerant {
 
-  def compare(l: UTF8String, r: UTF8String): Boolean
+  def compare(l: Array[Byte], r: Array[Byte]): Boolean
 
-  override def inputTypes: Seq[DataType] = Seq(StringType, StringType)
+  override def inputTypes: Seq[AbstractDataType] =
+    Seq(TypeCollection(StringType, BinaryType), TypeCollection(StringType, BinaryType))
+
+  def convertData(dataType: DataType): Any => Array[Byte] = dataType match {
+    case BinaryType => (input: Any) => input.asInstanceOf[Array[Byte]]
+    case StringType => (input: Any) => input.asInstanceOf[UTF8String].getBytes
```

Comment:
Not sure that handling of the string and binary types in `BinaryPredicate` is good approach. @cloud-fan @HyukjinKwon WDYT?

Reply:
Not sure here we should wrap `byte[]` to `UTF8String` or convert `UTF8String` to `byte[]`.
Of we should not support compare `byte[]` with `UTF8String`?

Reply:
@AngersZhuuuu do you know if other DBMSes support this?

Reply:
> @AngersZhuuuu do you know if other DBMSes support this?

https://docs.oracle.com/en/database/other-databases/nosql-database/21.2/sqlreferencefornosql/ends_with_function.html Oracle  will cast input as string type

Reply:
Maybe we can follow Oracle to convert binary data to `UTF8String`